### PR TITLE
Sync data from Find

### PR DESCRIPTION
**Pull request by tvararu**
### Context

We need to pull in providers, courses, and sites from Find.

### Changes proposed in this pull request

Create a support page at `/support/providers` that has a big green button that triggers a service that pulls in and massages the data.

To pull in the data, I've created the `FindAPI::Provider` and `FindAPI::Site` models, which use `json_api_client` underneath.

### Guidance to review

This is prototype-level code; but it does work, and maybe it's sufficient for MVP.

The `SyncProvidersFromFind` service is VERY brittle looking, but it's idempotent and should not cause a loss of data. We could conceivably merge it in, knowing very well it's technical debt (but contained to just one file), with a view to rethink it later when we implement the real syncing mechanism described in the ADR.

There's also only one basic end to end test covering this currently.

### Screenshots

Screenshot 2019-11-01 at 17 14 54

### Link to Trello card

[196 - Selecting a course](https://trello.com/c/HaxxM4vH/196-selecting-a-course-🏈-kickoff)

### CODE REVIEW

**Comment by duncanjbrown** [comment]
Couple of notes from a first look

- imo the test coverage isn't quite good enough to include this, as it doesn't cover the fiddly bit, ie the creation of Courses and CourseOptions. This could be easily fixed
- the implementation really covers fetching courses on a per-provider basis rather than for all providers so there's some value in renaming
- the UI looks good enough

**Comment by duncanjbrown** [comment]
ci failing for rubocop

File: config/routes.rb [hunk 1]
```diff
@@ -112,6 +112,9 @@
 
     get '/vendors' => 'manage_vendors#index'
     post '/vendors' => 'manage_vendors#create'
+
+    get '/providers' => 'providers#index', as: :providers
+    post '/providers' => 'providers#sync'
```
Comment (alexdesi) [inline config/routes.rb]: at first look `post '/providers'`  suggests that a new provider is created, but actually it is a sync, 
what do you think about changing the route to  '/providers/sync` ?

File: app/models/find_api/provider.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+module FindAPI
+  class Provider < FindAPI::Resource
+    RECRUITMENT_CYCLE_YEAR = ENV.fetch('RECRUITMENT_CYCLE_YEAR') { 2020 }
```
Comment (alexdesi) [inline app/models/find_api/provider.rb]: if you move the constant up to the `FindApi` module, we can remove the duplication
`RECRUITMENT_CYCLE_YEAR = ENV.fetch('RECRUITMENT_CYCLE_YEAR') { 2020 }`

File: app/controllers/support_interface/providers_controller.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+module SupportInterface
+  class ProvidersController < SupportInterfaceController
+    def index
+      @providers = Provider.all
+    end
+
+    def sync
+      Rails.configuration.providers_to_sync[:codes].each do |code|
+        SyncProviderFromFind.call(provider_code: code)
+      end
```
Comment (alexdesi) [inline app/controllers/support_interface/providers_controller.rb]: (I know you like this here, but this really sound like a service :) )

File: app/services/sync_provider_from_find.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    provider = Provider.find_or_create_by(code: provider_code)
+    provider.name = find_provider.provider_name
+    provider.save
+
+    sites = find_sites.map do |find_site|
+      site = provider.sites.find_or_create_by(code: find_site.location_name)
+      site.name = find_site.location_name
+      site.save
+      site
+    end
```
Comment (alexdesi) [inline app/services/sync_provider_from_find.rb]: This is complex to follow (for me at least :P), the map should just return an array, it'd be better to split this into a `.each` and then a simple `map`, if you could use private method for this would be great.
(I know it is a service, we could use `class << self` if we need more methods)

File: app/services/sync_provider_from_find.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      find_course.sites.each do |find_course_site|
+        CourseOption.find_or_create_by(
+          site_id: sites.find { |s| s.name == find_course_site.location_name }.id,
+          course_id: course.id,
+          vacancy_status: 'B', # TODO: Should this be reflected by `find_course.has_vacancies?`
+        )
+      end
+    end
+  end
+end
```
Comment (alexdesi) [inline app/services/sync_provider_from_find.rb]: this method is a bit hard to follow and to maintain, it would be nice to simplify it with some private and well named methods.

File: config/routes.rb (hunk 1 above)
Comment (duncanjbrown) [reply config/routes.rb]: Yes, much better :+1:

File: app/models/find_api/provider.rb (hunk 2 above)
Comment (duncanjbrown) [reply app/models/find_api/provider.rb]: Sounds like a good idea, will try this

File: app/services/sync_provider_from_find.rb (hunk 4 above)
Comment (duncanjbrown) [reply app/services/sync_provider_from_find.rb]: :+1: I'll see if I can clean this up a bit

**Comment by duncanjbrown** [comment]
Thanks for comments @alexdesi and @stevehook, addressing.

File: app/controllers/support_interface/providers_controller.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+module SupportInterface
+  class ProvidersController < SupportInterfaceController
+    def index
+      @providers = Provider.all
```
[the remark this answers is not in the document]
Comment (duncanjbrown) [reply app/controllers/support_interface/providers_controller.rb]: Good spot, fixed

File: app/controllers/support_interface/providers_controller.rb (hunk 3 above)
Comment (duncanjbrown) [reply app/controllers/support_interface/providers_controller.rb]: I prefer this interface as there is less reason for the service to change, and it can be easily reused by other callers

File: app/services/sync_provider_from_find.rb (hunk 4 above)
Comment (alexdesi) [reply app/services/sync_provider_from_find.rb]: Much better now! thanks
as I said, if you want you can group the `self.____` methods using the pattern `class << self`, and you can make `private` on those methods that you do not need to expose.
